plugin-server/src/ingestion/ingestion-consumer.ts

```diff
@@ -179,6 +178,38 @@ export class IngestionConsumer {
 
     public async handleKafkaBatch(messages: Message[]) {
         const parsedMessages = await this.runInstrumented('parseKafkaMessages', () => this.parseKafkaBatch(messages))
+        // Check if hogwatcher should be used (using the same sampling logic as in the transformer)
+        const shouldRunHogWatcher = Math.random() < this.hub.CDP_HOG_WATCHER_SAMPLE_RATE
+
+        // Get hog function IDs for all teams and cache function states only if hogwatcher is enabled
+        if (shouldRunHogWatcher) {
+            // Extract all team IDs from the batch of events
+            const teamIds = new Set<number>()
+            Object.values(parsedMessages).forEach((items) => {
+                items.forEach(({ event }) => {
+                    if (event.team_id) {
+                        teamIds.add(event.team_id)
+                    }
+                })
+            })
+
+            await this.runInstrumented('fetchAndCacheHogFunctionStates', async () => {
+                const teamIdsArray = Array.from(teamIds)
+                // Get hog function IDs for transformations
+                const teamHogFunctionIds = await this.hogTransformer['hogFunctionManager'].getHogFunctionIdsForTeams(
+                    teamIdsArray,
+                    ['transformation']
+                )
+
+                // Flatten all hog function IDs into a single array
+                const allHogFunctionIds = Object.values(teamHogFunctionIds).flat()
+
+                if (allHogFunctionIds.length > 0) {
+                    // Cache the hog function states
+                    await this.hogTransformer.fetchAndCacheHogFunctionStates(allHogFunctionIds)
+                }
+            })
+        }
```

Comment:
Nit: Would be nice to move this to its own private function. In addition - the "clearHogFunctionStates" could be done in that function first